psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1425 @@
+    set_scalar_variable("CURRENT ENERGY", e_mp2_total);
+
+    set_scalar_variable("MP2 SINGLES ENERGY", 0.0);
+    set_scalar_variable("MP2 DOUBLES ENERGY", e_mp2_corr);
+    set_scalar_variable("MP2 SAME-SPIN CORRELATION ENERGY", 0.0);      // TODO
+    set_scalar_variable("MP2 OPPOSITE-SPIN CORRELATION ENERGY", 0.0);  // TODO
+
+    return e_mp2_total;  // correct, or return just correlation energy?
+}
+
+void DLPNOMP2::print_header() {
+    outfile->Printf("   --------------------------------------------\n");
+    outfile->Printf("                     DLPNO-MP2                 \n");
+    outfile->Printf("                   by Zach Glick               \n");
+    outfile->Printf("   --------------------------------------------\n\n");
+    outfile->Printf("  DLPNO convergence set to %s.\n\n", options_.get_str("PNO_CONVERGENCE").c_str());
+    outfile->Printf("  Detailed DLPNO thresholds and cutoffs:\n");
+    outfile->Printf("    T_CUT_DO     = %6.3e \n", T_CUT_DO_);
+    outfile->Printf("    T_CUT_PNO    = %6.3e \n", T_CUT_PNO_);
+    outfile->Printf("    T_CUT_DO_ij  = %6.3e \n", options_.get_double("T_CUT_DO_ij"));
+    outfile->Printf("    T_CUT_PRE    = %6.3e \n", options_.get_double("T_CUT_PRE"));
+    outfile->Printf("    T_CUT_DO_PRE = %6.3e \n", options_.get_double("T_CUT_DO_PRE"));
+    outfile->Printf("    T_CUT_MKN    = %6.3e \n", options_.get_double("T_CUT_MKN"));
+    outfile->Printf("    T_CUT_CLMO   = %6.3e \n", options_.get_double("T_CUT_CLMO"));
+    outfile->Printf("    T_CUT_CPAO   = %6.3e \n", options_.get_double("T_CUT_CPAO"));
+    outfile->Printf("    S_CUT        = %6.3e \n", options_.get_double("S_CUT"));
+    outfile->Printf("    F_CUT        = %6.3e \n", options_.get_double("F_CUT"));
+    outfile->Printf("\n");
+}
+
+void DLPNOMP2::print_aux_domains() {
+    size_t total_atoms = 0, min_atoms = lmo_to_riatoms_[0].size(), max_atoms = 0;
+    for (auto atom_list : lmo_to_riatoms_) {
+        total_atoms += atom_list.size();
+        min_atoms = std::min(min_atoms, atom_list.size());
+        max_atoms = std::max(max_atoms, atom_list.size());
+    }
+
+    size_t total_bfs = 0, min_bfs = lmo_to_ribfs_[0].size(), max_bfs = 0;
+    for (auto bf_list : lmo_to_ribfs_) {
+        total_bfs += bf_list.size();
+        min_bfs = std::min(min_bfs, bf_list.size());
+        max_bfs = std::max(max_bfs, bf_list.size());
+    }
+
+    size_t naocc = lmo_to_ribfs_.size();
+    outfile->Printf("\n");
+    outfile->Printf("    Auxiliary BFs per Local MO:\n");
+    outfile->Printf("      Average = %4d AUX BFs (%d atoms)\n", total_bfs / naocc, total_atoms / naocc);
+    outfile->Printf("      Min     = %4d AUX BFs (%d atoms)\n", min_bfs, min_atoms);
+    outfile->Printf("      Max     = %4d AUX BFs (%d atoms)\n", max_bfs, max_atoms);
+}
+
+void DLPNOMP2::print_pao_domains() {
+    size_t total_atoms = 0, min_atoms = lmo_to_paoatoms_[0].size(), max_atoms = 0;
+    for (auto atom_list : lmo_to_paoatoms_) {
```

Comment:
There are a number of `auto` loops like this where the loop variable is read-only.  IMHO it's better to declare them like `const auto &atom_list` when that's the case a) to clearly convey the read-only nature and b) to give the compiler a chance to avoid copies if it's a large object.  I don't think it's necessary to change any of the loops in here, but just wanted to weigh in with that opinion for future reference.

Reply:
This is a very good point, thanks. Since some of those lists may be very long, it was definitely worth making your suggested change.